ntoskrnl/ke/i386/mproc.c

```diff
@@ -37,18 +38,37 @@ VOID
 NTAPI
 KeStartAllProcessors(VOID)
 {
-    PVOID KernelStack, DPCStack;
-    SIZE_T ProcessorCount = 0;
     PAPINFO APInfo;
+    PVOID KernelStack, DPCStack;
+    ULONG ProcessorCount;
+    ULONG MaximumProcessors;
+
+    // NOTE: NT6+ HAL exports HalEnumerateProcessors() and
+    // HalQueryMaximumProcessorCount() that help determining
+    // the number of detected processors on the system.
+    if (!KeMaximumProcessors)
+        KeMaximumProcessors = MAXIMUM_PROCESSORS;
+
+    MaximumProcessors = KeMaximumProcessors;
 
-    while (TRUE)
+    /* Limit the number of processors we can start at run-time */
+    if (KeNumprocSpecified)
+        MaximumProcessors = min(MaximumProcessors, KeNumprocSpecified);
+
+    /* Limit also the number of processors we can start during boot-time */
+    if (KeBootprocSpecified)
+        MaximumProcessors = min(MaximumProcessors, KeBootprocSpecified);
+
+    // TODO: Support processor nodes
+
+    /* Start ProcessorCount at 1 because we already have the boot CPU */
+    for (ProcessorCount = 1; ProcessorCount < MaximumProcessors; ++ProcessorCount)
```

Comment:
So we instead of starting all possible processors in the system and throwing the "Not permitted" ones in a loop. We're only sending the startup sequences to processors that are permitted to boot?

How do we deem what is the better CPUs to throw into execution if they're permitted. 
I wonder like SMT for instance, SMT once we support it is represented as other ""Cores"" i.e specifically they have they're own LAPICs to dispatch IPIs too.

For sure out of scope for this and I don't really mind, but it's something to think about. Maybe we should only fill real cores first until our "MaximumProcessors" are filled.

Reply:
Yes clearly I was not looking at the SMT thing so code will have to be somewhat improved I suppose, too... but existing code doesn't support them either for now.
Here we mostly care about logical processors, aren't we?
The loop here is meant to say: "start maximum as many as `MaximumProcessors`" -- if the boot switches are not given, the MaximumProcessors is the max number we can describe/support by the kernel (currently we just have one global KAFFINITY mask (a 32-bit value) to describe all of them, and a Prcb array of MAXIMUM_PROCESSORS == 32 entries)

Side question: where are we throwing the "Not permitted" ones in a loop?

Reply:
well we wouldn’t be yet anyway, I was thinking of starting every available processor and then at the increment of the KeActiveProcessorCount check if it’s higher then affinity on 32Bit builds and throwing them infinitely in a loop

Reply:
> So we instead of starting all possible processors in the system and throwing the "Not permitted" ones in a loop. We're only sending the startup sequences to processors that are permitted to boot?

Yeah I'm pretty sure we should avoid starting them at all.
I know a piece of system software that supports a "dedicated mode" where you tell Windows to use N/M processors, then their driver takes over the remaining CPUs to run its own scheduler.

Reply:
neat alright sounds good to me

Reply:
Oh, not that you need more convincing, but: I've also used /NUMPROC to disable defective CPU cores. So if they're still running code that kind of defeats the purpose ;)

Reply:
> Yeah I'm pretty sure we should avoid starting them at all. I know a piece of system software that supports a "dedicated mode" where you tell Windows to only use N out of M processors, then their driver takes over the remaining (M-N) CPUs to run its own scheduler.

This sounds like something the ["Windows Real-Time OS (RTX64)" software](https://www.intervalzero.com/en-products/en-rtx64/) would do :joy:

Reply:
It's not that one, but a similar software, yes